### PR TITLE
compact lock cf periodically

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Reduce lock cf's delete tags

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@BusyJay @siddontang @ngaut @disksing PTAL

**Comment by siddontang** [comment]
CI failed.

File: src/raftstore/store/config.rs [hunk 1]
```diff
@@ -36,6 +36,7 @@ const DEFAULT_MGR_GC_TICK_INTERVAL_MS: u64 = 60000;
 const DEFAULT_SNAP_GC_TIMEOUT_SECS: u64 = 60 * 10;
 const DEFAULT_MESSAGES_PER_TICK: usize = 256;
 const DEFAULT_MAX_PEER_DOWN_SECS: u64 = 300;
+const DEFAULT_LOCK_CF_COMPACT_INTERVAL_MS: u64 = 10 * 60 * 1000; // 10 min
```
Comment (siddontang) [inline src/raftstore/store/config.rs]: we can use seconds here, this operation is not frequent.

**Comment by siddontang** [comment]
Please add test to cover the operation.

File: src/raftstore/store/worker/compact.rs [hunk 2]
```diff
@@ -11,40 +11,43 @@
...
-    pub fn new(ps: &PeerStorage, compact_idx: u64) -> Task {
-        Task {
-            engine: ps.get_engine().clone(),
-            region_id: ps.get_region_id(),
-            compact_idx: compact_idx,
-        }
-    }
+pub enum Task {
+    CompactLockCF {
+        engine: Arc<DB>,
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: I think we can create the engine field in Runner, no need to pass it every time in Task.

File: src/raftstore/store/worker/compact.rs [hunk 3]
```diff
@@ -11,40 +11,43 @@
...
+        start_key: Option<Vec<u8>>, // None means smallest key
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: I think we don't need start and end key. Do we need to support compacting special range for the cf lock?

File: src/raftstore/store/worker/compact.rs [hunk 4]
```diff
@@ -15,16 +15,17 @@ use raftstore::store::keys;
...
 
 use rocksdb::{DB, WriteBatch, Writable};
 use std::sync::Arc;
 use std::fmt::{self, Formatter, Display};
 use std::error;
 
 pub enum Task {
-    CompactLockCF {
+    CompactRangeForCF {
         engine: Arc<DB>,
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: I still think we don't need to pass engine here, we can create the Runner with the engine and then use it.

File: src/raftstore/store/worker/compact.rs [hunk 5]
```diff
@@ -112,14 +118,14 @@ impl Runnable<Task> for Runner {
                 debug!("[region {}] execute compacting log to {}",
                        region_id,
                        compact_idx);
-                match self.compact_raft_log(engine.clone(), region_id, compact_idx) {
+                match self.compact_raft_log(engine, region_id, compact_idx) {
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: log compact time here.  or use a histogram metric.

**Comment by siddontang** [comment]
I think if we can add a threshold later, if the total deleted lock keys > threshold, we can trigger the compaction too.

File: src/raftstore/store/worker/compact.rs [hunk 6]
```diff
@@ -21,30 +21,38 @@ use rocksdb::{DB, WriteBatch, Writable};
...
-impl Task {
-    pub fn new(ps: &PeerStorage, compact_idx: u64) -> Task {
-        Task {
-            engine: ps.get_engine().clone(),
-            region_id: ps.get_region_id(),
-            compact_idx: compact_idx,
-        }
-    }
+pub enum Task {
+    CompactRangeForCF {
```
Comment (siddontang) [inline src/raftstore/store/worker/compact.rs]: I prefer CompactRangeCF

**Comment by siddontang** [comment]
If in the interval (10m), we only have little delete locks operations, do we need to trigger compaction?

**Comment by zhangjinpeng1987** [comment]
`If in the interval (10m), we only have little delete locks operations, do we need to trigger compaction?`
Compact range for lock cf is a trivial operation because the amount data of lock cf is mostly below several or dozens of MB, at the same time the compact frequency is limited at one time per 10 minutes.

**Comment by zhangjinpeng1987** [comment]
`I think if we can add a threshold later, if the total deleted lock keys > threshold, we can trigger the compaction too.`
This is a good idea, but I think compact range periodically also can fulfill this task and is very simple. Maybe we can do this after we add statistics(as guidance for GC) modular for scheduler.

**Comment by siddontang** [comment]
I think we should reduce the config time to test it.

**Comment by zhangjinpeng1987** [comment]
`I think we should reduce the config time to test it.`
Is it test the time cost by compact task or others?

File: src/raftstore/store/worker/metrics.rs [hunk 7]
```diff
@@ -62,4 +62,11 @@ lazy_static! {
             "Bucketed histogram of raftstore split check duration",
             labels!{"type" => "cost",}
         ).unwrap();
+
+    pub static ref COMPACT_RANGE_FOR_CF: HistogramVec =
+        register_histogram_vec!(
+            "tikv_compact_range_for_cf_duration_seconds",
+            "Bucketed histogram of compact range for cf execution",
+            &["type"]
```
Comment (disksing) [inline src/raftstore/store/worker/metrics.rs]: I think use 'cf' as label name is more clear.

**Comment by zhangjinpeng1987** [comment]
I reduce the manual compaction interval to 1 minute and trace the compaction task. Here are some logs about manual compaction.
tikv's log:
```
2016-10-08 10:46:13,445 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 108975 }
2016-10-08 10:47:13,445 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 99231 }
2016-10-08 10:48:13,644 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 62543 }
2016-10-08 10:49:13,645 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 93541 }
2016-10-08 10:50:13,769 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 24995100 }
2016-10-08 10:51:13,771 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 26336319 }
2016-10-08 10:52:13,760 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 15339832 }
2016-10-08 10:53:13,767 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 22528965 }
2016-10-08 10:54:13,768 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 22362935 }
2016-10-08 10:55:13,972 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 27311961 }
2016-10-08 10:56:13,974 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 29464734 }
2016-10-08 10:57:13,965 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 20317008 }
2016-10-08 10:58:13,964 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 19505623 }
2016-10-08 10:59:13,963 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 18111600 }
2016-10-08 11:00:13,967 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 22122679 }
2016-10-08 11:01:13,993 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 47772900 }
2016-10-08 11:02:13,973 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 27795037 }
2016-10-08 11:03:14,159 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 14596270 }
2016-10-08 11:04:14,158 compact.rs:123 - INFO  - compact range for lock cf take duration Duration { secs: 0, nanos: 13143089 }
```
rocksdb's log
```
2016/10/08-10:46:13.445716 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:47:13.445777 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:48:13.644795 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:49:13.645349 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:50:13.763073 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:50:13.763149 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:50:13.769802 7fcd329ff700 (Original Log Time 2016/10/08-10:50:13.763206) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-10:51:13.763868 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:51:13.765551 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:51:13.765797 7fcd315ff700 (Original Log Time 2016/10/08-10:51:13.765722) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-10:52:13.757778 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:52:13.760208 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:53:13.760851 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:53:13.761725 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:53:13.761866 7fcd33dff700 (Original Log Time 2016/10/08-10:53:13.761827) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-10:54:13.763864 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:54:13.764984 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:54:13.767917 7fcd335fe700 (Original Log Time 2016/10/08-10:54:13.765120) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-10:55:13.961336 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:55:13.962199 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:55:13.962365 7fcd309ff700 (Original Log Time 2016/10/08-10:55:13.962333) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-10:56:13.972491 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:56:13.974753 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:57:13.955727 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:57:13.965394 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:58:13.961700 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:58:13.962104 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-10:58:13.964652 7fcd335fe700 (Original Log Time 2016/10/08-10:58:13.962139) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-10:59:13.959281 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-10:59:13.963569 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-11:00:13.964794 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-11:00:13.965208 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-11:00:13.967587 7fcd321fe700 (Original Log Time 2016/10/08-11:00:13.965280) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-11:01:13.986602 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-11:01:13.987696 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-11:01:13.987846 7fcd329ff700 (Original Log Time 2016/10/08-11:01:13.987816) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-11:02:13.966357 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-11:02:13.973553 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-11:03:14.155614 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-11:03:14.155851 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-11:03:14.155922 7fcd309ff700 (Original Log Time 2016/10/08-11:03:14.155905) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
2016/10/08-11:04:14.156061 7fcd2c9fd700 [lock] Manual compaction waiting for all other scheduled background compactions to finish
2016/10/08-11:04:14.156287 7fcd2c9fd700 [lock] Manual compaction starting
2016/10/08-11:04:14.158619 7fcd33dff700 (Original Log Time 2016/10/08-11:04:14.156337) [lock] Manual compaction from level-0 to level-1 from (begin) .. (end); will stop at (end)
```
As we can see, the compact lock cf task takes about several milliseconds, and waiting for all other scheduled bg compactions to finish take the most of time, the time actually cost by compaction of lock cf is trivial.

**Comment by siddontang** [comment]
CI failed @BusyJay

File: src/raftstore/store/worker/metrics.rs [hunk 8]
```diff
@@ -62,4 +62,11 @@ lazy_static! {
             "Bucketed histogram of raftstore split check duration",
             labels!{"type" => "cost",}
         ).unwrap();
+
+    pub static ref COMPACT_RANGE_CF: HistogramVec =
+        register_histogram_vec!(
+            "tikv_compact_range_for_cf_duration_seconds",
```
Comment (siddontang) [inline src/raftstore/store/worker/metrics.rs]: s/for_cf/cf/

**Comment by siddontang** [comment]
Rest LGTM

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**